contents/code/tilingmanager.js

```diff
@@ -976,3 +1006,15 @@ TilingManager.prototype._getLayouts = function(desktop, screen) {
     }
     return null;
 };
+
+TilingManager.prototype._isDesktopEmpty = function(desktop) {
+    var clients = workspace.clientList();
+    var empty = true;
+    for (var i = 0; i < clients.length; i++) {
+        if (!clients[i].onAllDesktops && clients[i].desktop == desktop) {
```

Comment:
So clients on all desktops don't count?

That's probably the right decision, otherwise this would do nothing if you had any.

Should it only count tiled clients?

Otherwise this should work as a standalone script.

Reply:
> So clients on all desktops don't count?
That's probably the right decision, otherwise this would do nothing if you had any.

Yep, that's a good question. I did think about it and my conclusion was that it's a matter of taste and I prefer to not count sticky clients :wink: 

> Should it only count tiled clients?

Why?

> Otherwise this should work as a standalone script.

Well, technically you are right. This commit is in fact a complement to [another change](https://github.com/laloch/kwin-tiling/commit/9d32f6bb24b182735361a3cdb0a623069d6f33b3) I have in my kitchen. Once I have the `delete-desktop` branch ready (it unfortunately does not yet work too well), it makes perfect sense to have them both.

Reply:
I can, of course, join the two commits into one PR once the other half is ready and you can reject both the PRs later :wink:
Btw, would you mind opening a project chat over at https://gitter.im ? I have an unrelated question I'd like to ask.

Reply:
>>Should it only count tiled clients?

>Why?

Well, I guess that depends on what this is for, and how you view tiled and floating clients.

If you think that floating clients aren't considered at all by the script, it makes some sense to not count them in this one specific place.

>> Otherwise this should work as a standalone script.

> Well, technically you are right. This commit is in fact a complement to another change I have in my kitchen. Once I have the delete-desktop branch ready (it unfortunately does not yet work too well), it makes perfect sense to have them both.

What if that were _also_ in the other script?

What's the plan here?

I mean there's always this trade-off between integration and loose coupling, but all of this sounds like it'd make sense even without tiling, and like it wouldn't need to be integrated with tiling to work with it.

Though obviously that means tiling users by default wouldn't get those features.

One example where I've erred on the side of loose coupling is my own "ministack" script - https://github.com/faho/kwin-ministack. It works quite nicely with tiling, but since it's fully functional without I decided to keep it as a separate script.

I guess I'm asking one of two questions:

- What does this gain from being part of this script?
- What does the script gain by having this be a part of it?

Reply:
>Btw, would you mind opening a project chat over at https://gitter.im ? I have an unrelated question I'd like to ask.

https://gitter.im/kwin-tiling/community#

Reply:
> What does this gain from being part of this script?

Unfortunately, it cannot work as a standalone script. There's AFAIK no way to remove a desktop from the middle of the `desktops` array - you can only remove the last one. Before removal of a desktop from the middle I first need to "compact" all the populated desktops in such a way that the empty desktop remains in the last position and remove it afterwards. If I want to keep the layouts as they were on the compacted desktops (and I do want it), this code must be integrated with the tiling script.

> What does the script gain by having this be a part of it?

At least for me this is a very cool feature. The `delete-desktop` branch does not work well so far and yet I did already get used to the automatic desktop creation/destruction. It can be a bit confusing, because there are no fixed desktop numbers any more (that's why it is disabled by default), but I still find it very convenient to use. Try for yourself ;)

Reply:
>Unfortunately, it cannot work as a standalone script. There's AFAIK no way to remove a desktop from the middle of the desktops array - you can only remove the last one

Soo... moving apps to an empty desktop could "always" use a new one at the end, and that could always be removed once the app exits.

I imagine the fear here is that you'll get a ton of empty desktops in the middle?

Or am I missing another reason why you'd have to remove desktops in the middle?

>At least for me this is a very cool feature.

Oh, don't get me wrong, it _is_!

It's just that more people could use it if it worked without tiling!

Reply:
> I imagine the fear here is that you'll get a ton of empty desktops in the middle?

Yep, exactly. I have now the `_compactDesktops` call commented out, because it doesn't work well (I will need to completely suspend tiling while I'm moving the whole layouts), so yes, there can be few empty desktops in the middle.  That's why the Meta+D handler searches for an existing empty desktop first.